### PR TITLE
modernize crate

### PR DESCRIPTION
**Pull request by fogti**
- bump rust edition to 2018
- remove an unnecessary `Debug` constraint
- provide an `size_hint` implementation
- get rid of unnecessary lifetime annotations
- make crate `no_std`

... etc. see commit description.

### CODE REVIEW

File: Cargo.toml
Comment (fitzgen) [inline Cargo.toml]: Let's just jump to 1.0.0; I think the API is pretty stable at this point.

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs]: I don't think we need to build fusing into this iterator. We can just defer to the nested iterator, which can be fused if necessary. Otherwise we risk redundant fusing.

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs]: And without the `self.flag` fusing, then this method becomes simply

```suggestion
        // Can't know a lower bound, due to the predicate.
        (0, self.iter.size_hint().1)
```

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
        while let Some(x) = self.iter.next_if(&mut self.predicate) {
            accum = f(accum, x);
        }
        accum
```

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs]: And then let's change this to be fused only if the nested iterator is fused.

**Review by fitzgen — CHANGES_REQUESTED** [review]
Thanks for sending this PR! A couple changes requested below, and once they're addressed we can merge and publish.

File: src/lib.rs
Comment (fogti) [reply src/lib.rs]: huh, that's already the case.

File: src/lib.rs
Comment (fogti) [reply src/lib.rs]: I pretty much took it from the std `TakeWhile` implementation, and I realized why it makes sense: It is necessary to guarantee that the closure is not executed again and again when checking if the iterator would return `None`, which would also happen when "extended" with further iterators. It can be a bit surprising if this is not done. It might be a good idea to simply provide two implementations: A simple one which isn't fuse, but may call the closure more often than necessary, which may lead to surprises if the closure has side-effects, and a more complex one which is fused.

File: src/lib.rs
Comment (fitzgen) [reply src/lib.rs]: Oops, wasn't reading closely enough here, nvm about this comment.

File: src/lib.rs
Comment (fitzgen) [reply src/lib.rs]: But the closure isn't called unless the underlying iterator returns `Some`, so I don't see how this would be more often than necessary.

Comment (fogti) [reply src/lib.rs]: yes, but because it is a `take_while` implementation, the underlying iterator could return `Some` (causing the `peek()` to pick that up), but the `next_if` invocation doesn't forward it, because the predicate returns `false`. On the next call to `next()` of the `take_while` iterator, the predicate would be called again, because the underlying `Peekable` instance still holds the same element, as an unconditional `next()` on the `Peekable` instance was never called.

File: src/lib.rs
Comment (fogti) [reply src/lib.rs]: I think the implementation without `flag` would in fact not satisfy the `FusedIterator` contract, because the `predicate` may do weird stuff... And the implementation with `flat` would always satisfy the `FusedIterator` contract, because it explicitly fuses it...

**Comment by fogti** [comment]
ok, I think two implementations might be unnecessary, as calling `.fuse()` should be enough to get the expected behavoir.

File: src/lib.rs
Comment (fogti) [reply src/lib.rs]: I added the `not_fused` test case, which demonstrates it.

File: src/lib.rs
Comment (fitzgen) [inline src/lib.rs:203]: Ah good catch!

**Review by fitzgen — APPROVED** [review]
Thanks! Looks great!

**Pull request merged**
